### ISSUE TITLE
Pep proxy is not able to contact the IdM

### ISSUE THREAD

**Issue by tmontanaro** [FIWARE/tutorials.XACML-Access-Rules#6]
If I run the tutorial through the script "services" (docker-compose version: 1.24.0, build 0aa59064) and I try to contact the Orion context Broker through the Pep Proxy (and so through the APIs) I obtain the error `Error in IDM communication`. And it is related to problems experienced by the Pep proxy in contacting the Idm. In fact, if I look into the logs of the `fiware-orion-proxy` container, I find the following:
```
2019-05-15 09:38:13.348  - INFO: Server - Starting PEP proxy in port 1027. IdM authentication ...
2019-05-15 09:38:13.387  - ERROR: Server - Error in IDM communication Error: connect ECONNREFUSED 172.18.1.5:3005
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1191:14)
2019-05-15 09:38:13.387  - INFO: Server - Waiting 5 seconds before attempting again.
2019-05-15 09:38:18.396  - ERROR: Server - Error in IDM communication Error: connect ECONNREFUSED 172.18.1.5:3005
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1191:14)
2019-05-15 09:38:18.397  - INFO: Server - Waiting 5 seconds before attempting again.
2019-05-15 09:38:23.401  - ERROR: Server - Error in IDM communication Error: connect ECONNREFUSED 172.18.1.5:3005
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1191:14)
2019-05-15 09:38:23.401  - INFO: Server - Waiting 5 seconds before attempting again.
2019-05-15 09:38:28.405  - ERROR: Server - Error in IDM communication Error: connect ECONNREFUSED 172.18.1.5:3005
```

**Comment by jason-fox** [comment]
Hmmm weird (as you may guess I'm unable to reproduce the error myself). `ECONNREFUSED 172.18.1.5:3005` would indicate that **Keyrock** is not listening yet.

* What does the log of **Keyrock** say?  Can you log in to it?
* Is MySQL running properly?  Has the dummy data been injected?
* Have you recently pulled all containers using `./services create`?
* When you run `docker ps` are any containers `unhealthy`

**Comment by tmontanaro** [comment]
Thanks for the fast reply.
1. The last part of the Keyrock log in the following (after obtaining a new token and trying to send a GET request to the Orion Context Broker). It seems that the requests are redirected correctly but the response to the client is ```Error in IDM communication```.

```
Wed, 15 May 2019 12:03:20 GMT idm:oauth_controller  --> token
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server -------getClient-------
Executing (default): SELECT `id`, `redirect_uri`, `token_types`, `jwt_secret`, `scope`, `grant_type` FROM `oauth_client` AS `OauthClient` WHERE `OauthClient`.`id` = 'tutorial-dckr-site-0000-xpresswebapp' AND `OauthClient`.`secret` = 'tutorial-dckr-site-0000-clientsecret';
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server -------getIdentity-------
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server tutorial-dckr-site-0000-xpresswebapp
Executing (default): SELECT `id`, `username`, `gravatar`, `email`, `salt`, `password`, `scope`, `eidas_id`, `extra` FROM `user` AS `User` WHERE `User`.`email` = 'iof2020@fakeemail.com';
Executing (default): SELECT `id`, `password`, `salt` FROM `iot` AS `Iot` WHERE `Iot`.`id` = 'iof2020@fakeemail.com' AND `Iot`.`oauth_client_id` = 'tutorial-dckr-site-0000-xpresswebapp';
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server -------validateScope-------
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server -------saveToken-------
Wed, 15 May 2019 12:03:20 GMT idm:oauth2-model_oauth_server -------storeToken-------
Executing (54d8a6db-95bf-495f-b017-d04bd62a0161): START TRANSACTION;
Executing (default): INSERT INTO `oauth_refresh_token` (`refresh_token`,`expires`,`scope`,`valid`,`oauth_client_id`,`user_id`,`iot_id`,`authorization_code`) VALUES ('d9a8a2ac881d9d6dc33acfa6cf2df4d27cc04234','2019-05-29 12:03:20','bearer',true,'tutorial-dckr-site-0000-xpresswebapp','d629ebcd-4b54-4cc3-a34f-18e913fa3211',NULL,NULL);
Executing (54d8a6db-95bf-495f-b017-d04bd62a0161): SELECT `id`, `user_id`, `oauth_client_id` FROM `user_authorized_application` AS `User_Authorized_Application` WHERE `User_Authorized_Application`.`user_id` = 'd629ebcd-4b54-4cc3-a34f-18e913fa3211' AND `User_Authorized_Application`.`oauth_client_id` = 'tutorial-dckr-site-0000-xpresswebapp' LIMIT 1;
Executing (default): INSERT INTO `oauth_access_token` (`access_token`,`expires`,`scope`,`valid`,`oauth_client_id`,`user_id`,`iot_id`,`refresh_token`,`authorization_code`) VALUES ('3220288d1ecce82d41e20ff12a8788767ebe1f35','2019-05-15 13:03:20','bearer',true,'tutorial-dckr-site-0000-xpresswebapp','d629ebcd-4b54-4cc3-a34f-18e913fa3211',NULL,'d9a8a2ac881d9d6dc33acfa6cf2df4d27cc04234',NULL);
Executing (54d8a6db-95bf-495f-b017-d04bd62a0161): COMMIT;
POST /oauth2/token 200 172.837 ms - 177
GET /version 200 4.065 ms - 257
```
2. I think so, because if I connect to the GUI I can see alll the applications and all the users and rules. In any case, in the DB I can see both the "idm" database and the following tables:
   - SequelizeMeta
   - auth_token
   - authzforce
   - eidas_credentials
   - iot
   - oauth_access_token
   - oauth_authorization_code
   - oauth_client
   - oauth_refresh_token
   - oauth_scope
   - organization
   - pep_proxy
   - permission
   - role
   - role_assignment
   - role_permission
   - trusted_application
   - user
   - user_authorized_application
   - user_organization
   - user_registration_profile
3. Yes, in any case I always delete all the images and volumes before restarting new test
4. Everything is healthy (fiware-tutorial, fiware-iot-agent, fiware-keyrock, fiware-orion, fiware-authzforce) or running (fiware-orion-proxy, db-mongo, db-mysql)

**Comment by tmontanaro** [comment]
Wait: I restarted the pep proxy and now it is working (it did not help in previous trials)!
So maybe a delay in starting the pep proxy could help?

**Comment by tmontanaro** [comment]
I found the problem:
8080 port was also used by another service and thus it created conflicts!
Thus, I changed the port at which the Authzforce service is exposed 
```
    ports:
      - "8082:8080"
```
For other users: do not forget that the second port should not be changed because the server is run at port 8080 in the local network! In addition, all the other containers should use the 8080 port if they are run in the same network in which the Authzforce is run, while they should use 8082 if run in a different network and should contact the service through the Internet.
